File: Trees/Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (root == nullptr) {
            return false;
        }
        return helper(root, 0, targetSum);
    }

    bool helper(TreeNode* next, int prevSum, int targetSum) {
        int currSum = prevSum + next->val;

        if(next->right == nullptr && next->left == nullptr && currSum == targetSum) {
            return true;
        } else if (next->right == nullptr && next->left == nullptr && currSum != targetSum) {
            return false;
        } else if (next->left != nullptr && helper(next->left, currSum, targetSum)){
            return true;
        } else if (next->right != nullptr && helper(next->right, currSum, targetSum)){
            return true;
        } else {
            return false;
        }

    }
};

File: Trees/Path Sum.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        match root {
            Some(node) => Self::helper(&node, target_sum),
            None => false
        }
    }

    fn helper(node: &Rc<RefCell<TreeNode>>, target_sum: i32) -> bool {
        let node_ref = node.borrow();
        let current_sum = target_sum - node_ref.val;

        match (&node_ref.left, &node_ref.right) {
            (None, None) => {
                current_sum == 0
            }
            (Some(left), Some(right)) => {
                Self::helper(left, current_sum) || Self::helper(right, current_sum)
            }
            (Some(left), None) => {
                Self::helper(left, current_sum)
            }
            (None, Some(right)) => {
                Self::helper(right, current_sum)
            }
        }
    }
}